### ISSUE TITLE
deep filtering & deep finding

### ISSUE THREAD

**Issue by EtaiG** [lodash/lodash#591]
Not really an issue, just want some advice.

I often find that I need to perform some sort of deep filter, either with a known number of steps, or recursively.

For example, assuming the following schemas:

```
Collection: [ imageGroup ]  //array of imagegroups
imageGroup: {
title: 'string'
images: [image]  //array of images
}

image: {
   key1: 'string',
   key2: 'string',
   url: 'string',
   'width': 'number
}
```
1. If I want to get all of the images from the collection which match a certain filter, how should I do it? Is there a more elegant way than the following:

```
_.reduce(collection, function(accum, imageGroup){
   return accum.concat(_.filter(imageGroup, function(image){
   //filter logic here
}));
}, [])
```
1. If I want to do a recursive filter, which will return the actual element - something similar to:

```
function findRecurse(node){
 if (node.type === 'sometype')
return node;

var result;
for(var i=0;i<node.children.length;i++){
 result = findRecurse(node.children[i]);
if(result) return result;
}
return false;
}
```

in other words, is there a more appropriate way, using lodash, to recursively find the first 'deep' element that the callback returns truey for?

Thanks :)

**Comment by AndrewEastwood** [comment]
Hey @jdalton, I'd like to have smth like deepSearch function too.

**Comment by AndrewEastwood** [comment]
can you add this function into distro?

**Comment by jdalton** [comment]
@AndrewEastwood for a deep find see [lodash-deep](https://www.npmjs.com/package/lodash-deep).

**Comment by StreetStrider** [comment]
@EtaiG, what type of objects is in `[ images ]`, what must be «matched a certain filter»? Is it a binary? Or it is an object with some kind of tags?

**Comment by YuriGor** [comment]
Take a look at [_.eachDeep](https://github.com/YuriGor/deepdash#eachdeep-foreachdeep):
```js
var images = [];
_.eachDeep(imgCollection,function(value, key, path, depth, parent){
  if(key=="key1" && value=="take this")
    images.push(parent);
});
```
